Assets/Scripts/Gameplay/ConnectionManagement/ConnectionState/ConnectionState.cs

```diff
@@ -0,0 +1,83 @@
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Unity.Multiplayer.Samples.BossRoom.Shared.Infrastructure;
+using Unity.Netcode;
+using UnityEngine;
+
+namespace Unity.Multiplayer.Samples.BossRoom
+{
+    /// <summary>
+    /// Base class representing a connection state. Contains a static member for each possible state.
+    /// </summary>
+    public abstract class ConnectionState
+    {
+        public static readonly OfflineState Offline = new OfflineState();
+        public static readonly ClientConnectingState ClientConnecting = new ClientConnectingState();
+        public static readonly ClientConnectedState ClientConnected = new ClientConnectedState();
+        public static readonly ClientReconnectingState ClientReconnecting = new ClientReconnectingState();
+        public static readonly DisconnectingWithReasonState DisconnectingWithReason = new DisconnectingWithReasonState();
+        public static readonly HostingState Hosting = new HostingState();
+
+        public static readonly List<ConnectionState> States = new() { Offline, ClientConnecting, ClientConnected, ClientReconnecting, DisconnectingWithReason, Hosting };
+
+        /// <summary>
+        /// Initializes each of the static states, by setting the reference to the ConnectionManager and injecting dependencies.
+        /// </summary>
+        /// <param name="connectionManager"></param>
+        /// <param name="scope"></param>
+        public static void InitializeStates(ConnectionManager connectionManager, DIScope scope)
+        {
+            foreach (var connectionState in States)
+            {
+                connectionState.m_ConnectionManager = connectionManager;
+                scope.InjectIn(connectionState);
+            }
+        }
+
+        protected ConnectionManager m_ConnectionManager;
+
+        public abstract void Enter();
+
+        public abstract void Exit();
+
+        public virtual void OnClientConnected(ulong clientId) { }
+        public virtual void OnClientDisconnect(ulong clientId) { }
+
+        public virtual void StartClientIP(string playerName, string ipaddress, int port) { }
+
+        public virtual Task StartClientLobbyAsync(string playerName, Action<string> onFailure)
+        {
+            return Task.CompletedTask;
+        }
+
+        public virtual bool StartHostIP(string playerName, string ipaddress, int port)
+        {
+            return false;
+        }
+
+        public virtual Task StartHostLobbyAsync(string playerName)
+        {
+            return Task.CompletedTask;
+        }
+
+        public virtual void OnUserRequestedShutdown() { }
+
+        public virtual void OnDisconnectReasonReceived(ConnectStatus disconnectReason) { }
+
+        /// <summary>
```

Comment:
I could move the "remarks" section to the implementations, yes, but I'd keep the rest here at least, since there could be multiple implementations. Anyways, this will be changed if pre.10 is released before this PR is merged

Reply:
yeah, I changed the comment in my PR

Reply:
I mean, the logic it's describing is in the implementation, let's have them close to each other. A comment describing a method signature isn't super helpful...